gmerlin/apps/cmdlinerecorder/gmerlin-record.1
.TH GMERLIN-RECORD 1 "June 2012" Gmerlin "User Manuals"
.SH NAME
gmerlin-record \- Commandline recorder
.SH SYNOPSIS
.B gmerlin-record 
.I [options]

.SH DESCRIPTION
Commandline recorder

.SH OPTIONS

.B \-aud
.I <audio_options>

.RS 2
Set audio recording options
.RE
.RS 2
.BR do_audio "=[1|0] (default: 1)"
.P
Record audio
.P
.BR plugin "=option[{suboptions}]"
.P
Supported options: i_alsa i_oss i_esd i_pulse i_audiofile
.P
Default: i_alsa
.P
Plugin
.P
.B Suboptions for plugin=i_alsa

.P
.RS 2
.BR card "=<string>"
.P
Supported strings: default hw:0,0 hw:0,2
.P
Default: default
.P
Card
.P
.BR channel_mode "=<string>"
.P
Supported strings: mono stereo
.P
Default: stereo
.P
Channel Mode
.P
.BR bits "=<string>"
.P
Supported strings: 8 16
.P
Default: 16
.P
Bits
.P
.BR samplerate "=<number> (8000..96000, default: 44100)"
.P
Samplerate [Hz]
.P
.BR buffer_time "=<number> (10..10000, default: 1000)"
.P
Buffer time
.P
Set the buffer time (in milliseconds). Larger values improve recording
performance on slow systems under load.

.P
.BR user_device "=<string>"
.P
User device
.P
Enter a custom device to use for recording. Leave empty to use the
settings above

.P
.RE
.B Suboptions for plugin=i_oss

.P
.RS 2
.BR device "=<string> (Default: /dev/dsp)"
.P
Device
.P
.BR channel_mode "=<string>"
.P
Supported strings: Mono Stereo
.P
Default: Stereo
.P
Channel Mode
.P
.BR bits "=<string>"
.P
Supported strings: 8 16
.P
Default: 16
.P
Bits
.P
.BR samplerate "=<number> (8000..96000, default: 44100)"
.P
Samplerate [Hz]
.P
.RE
.B Suboptions for plugin=i_esd

.P
.RS 2
.BR esd_host "=<string>"
.P
Host (empty: local)
.P
.BR input_mode "=<string>"
.P
Supported strings: record monitor
.P
Default: record
.P
Input Mode
.P
.RE
.B Suboptions for plugin=i_pulse

.P
.RS 2
.BR channel_mode "=<string>"
.P
Supported strings: mono stereo
.P
Default: stereo
.P
Channel Mode
.P
.BR bits "=<string>"
.P
Supported strings: 8 16
.P
Default: 16
.P
Bits
.P
.BR samplerate "=<number> (8000..96000, default: 44100)"
.P
Samplerate [Hz]
.P
.BR server "=<string>"
.P
Server
.P
Server to connect to. Leave empty for default.

.P
.BR dev "=<string>"
.P
Source
.P
Source to open. Use the PulseAudio manager for available Sources.

.P
.RE
.B Suboptions for plugin=i_audiofile

.P
.RS 2
.BR channel_mode "=<string>"
.P
Supported strings: mono stereo
.P
Default: stereo
.P
Channel Mode
.P
.BR samplerate "=<number> (8000..96000, default: 44100)"
.P
Samplerate [Hz]
.P
.BR album "=<string>"
.P
Album
.P
.BR shuffle "=[1|0] (default: 0)"
.P
Shuffle
.P
.RE
.BR language "=<string>"
.P
Supported strings: aar abk ace ach ada afa afh afr aka akk ale alg amh ang apa ara arc arn arp art arw asm ath aus ava ave awa aym aze bad bai bak bal bam ban bas bat bej bel bem ben ber bho bih bik bin bis bla bnt tib bos bra bre btk bua bug bul cad cai car cat cau ceb cel cze cha chb che chg chk chm chn cho chp chr chu chv chy cmc cop cor cos cpe cpf cpp cre crp cus wel dak dan day del den ger dgr din div doi dra dua dum dyu dzo efi egy eka gre elx eng enm epo est baq ewe ewo fan fao per fat fij fin fiu fon fre frm fro fry ful fur gaa gay gba gem gez gil gla gle glg glv gmh goh gon gor got grb grc grn guj gwi hai hau haw heb her hil him hin hit hmn hmo scr hun hup arm iba ibo ijo iku ile ilo ina inc ind ine ipk ira iro ice ita jav jpn jpr kaa kab kac kal kam kan kar kas geo kau kaw kaz kha khi khm kho kik kin kir kmb kok kom kon kor kos kpe kro kru kum kur kut lad lah lam lao lat lav lez lin lit lol loz ltz lua lub lug lui lun luo lus mad mag mah mai mak mal man map mar mas mdr men mga mic min mis mac mkh mlg mlt mnc mni mno moh mol mon mos mao may mul mun mus mwr bur myn nah nai nau nav nbl nde ndo nds nep new nia nic niu dut nno nob non nor nso nub nya nym nyn nyo nzi oci oji ori orm osa oss ota oto paa pag pal pam pan pap pau peo phi pli pol pon por pra pro pus que raj rap rar roa rom rum run rus sad sag sah sai sal sam san sas sat sco sel sem sga sgn shn sid sin sio sit sla slo slv sme smi smo sna snd snk sog som son sot spa alb srd scc srr ssa ssw suk sun sus sux swa swe syr tah tai tam tat tel tem ter tet tgk tgl tha tig tir tiv tkl tli tmh tog ton tpi tsi tsn tso tuk tum tur tut tvl twi tyv uga uig ukr umb und urd uzb vai ven vie vol vot wak wal war was wen wol xho yao yap yid yor ypk zap zen zha chi znd zul zun
.P
Default: eng
.P
Language
.P
.RE

.B \-vid
.I <video_options>

.RS 2
Set video recording options
.RE
.RS 2
.BR do_video "=[1|0] (default: 1)"
.P
Record video
.P
.BR plugin "=option[{suboptions}]"
.P
Supported options: i_v4l2 i_v4l i_x11
.P
Default: i_v4l2
.P
Plugin
.P
.B Suboptions for plugin=i_v4l2

.P
.RS 2
.BR device "=option[{suboptions}]"
.P
Supported options: <None>
.P
Default: /dev/video0
.P
V4L2 Device
.P
.BR force_rw "=[1|0] (default: 1)"
.P
Force read
.P
Don't use memory mapping

.P
.BR resolution "=<string>"
.P
Supported strings: QSIF (160x112) QCIF (176x144) QVGA (320x240) SIF(352x240) CIF (352x288) VGA (640x480) User defined
.P
Default: QVGA (320x240)
.P
Resolution
.P
.BR user_width "=<number> (160..1024, default: 720)"
.P
User defined width
.P
.BR user_height "=<number> (112..768, default: 576)"
.P
User defined height
.P
.RE
.B Suboptions for plugin=i_v4l

.P
.RS 2
.BR device "=<string> (Default: /dev/video0)"
.P
V4L Device
.P
.BR resolution "=<string>"
.P
Supported strings: QSIF (160x112) QCIF (176x144) QVGA (320x240) SIF(352x240) CIF (352x288) VGA (640x480) User defined
.P
Default: QVGA (320x240)
.P
Resolution
.P
.BR user_width "=<number> (160..1024, default: 720)"
.P
User defined width
.P
.BR user_height "=<number> (112..768, default: 576)"
.P
User defined height
.P
.BR brightness "=<number> (0..65535, default: 40000)"
.P
Brightness
.P
.BR hue "=<number> (0..65535, default: 65535)"
.P
Hue
.P
.BR colour "=<number> (0..65535, default: 30000)"
.P
Colour
.P
.BR contrast "=<number> (0..65535, default: 30000)"
.P
Contrast
.P
.BR whiteness "=<number> (0..65535, default: 30000)"
.P
Whiteness
.P
.RE
.B Suboptions for plugin=i_x11

.P
.RS 2
.BR root "=[1|0] (default: 0)"
.P
Grab from root window
.P
.BR draw_cursor "=[1|0] (default: 1)"
.P
Draw cursor
.P
.BR win_ontop "=[1|0] (default: 1)"
.P
Keep grab window on top
.P
.BR win_sticky "=[1|0] (default: 1)"
.P
Make grab window sticky
.P
.BR disable_screensaver "=[1|0] (default: 1)"
.P
Disable screensaver
.P
Disable screensaver and energy saving mode

.P
.RE
.BR do_limit_fps "=[1|0] (default: 1)"
.P
Limit fps
.P
Limit frames per second. By default, the maximum possible framerate will be
used.

.P
.BR limit_fps "=<number> (1.00..100.00, default: 25.00)"
.P
fps limit
.P
Specify the maximum framerate for input plugins, which can capture images
really fast

.P
.RE

.B \-af
.I <audio filter options>

.RS 2
Set audio filter options
.RE
.RS 2
.BR audio_filters "={option[{suboptions}][:option[{suboptions}]...]}"
.P
Supported options: fa_ladspa_fourByFourPole fa_ladspa_amPitchshift 
fa_ladspa_alias fa_ladspa_allpass_c fa_ladspa_allpass_l 
fa_ladspa_allpass_n fa_ladspa_amp_mono fa_ladspa_amp_stereo 
fa_ladspa_artificialLatency fa_ladspa_divider fa_ladspa_autoPhaser 
fa_ladspa_satanMaximiser fa_ladspa_canyon_delay fa_ladspa_chebstortion 
fa_ladspa_comb fa_ladspa_comb_c fa_ladspa_comb_l fa_ladspa_comb_n 
fa_ladspa_const fa_ladspa_crossoverDist fa_ladspa_dcRemove fa_ladspa_dj_eq 
fa_ladspa_dj_eq_mono fa_ladspa_djFlanger fa_ladspa_decimator 
fa_ladspa_declip fa_ladspa_delayorama fa_ladspa_diode 
fa_ladspa_disintegrator fa_ladspa_dysonCompress fa_ladspa_delay_0,01s 
fa_ladspa_delay_0,1s fa_ladspa_delay_1s fa_ladspa_delay_5s 
fa_ladspa_delay_60s fa_ladspa_decay fa_ladspa_fmOsc 
fa_ladspa_fastLookaheadLimiter fa_ladspa_foverdrive 
fa_ladspa_fbdelay_0,01s fa_ladspa_fbdelay_0,1s fa_ladspa_fbdelay_1s 
fa_ladspa_fbdelay_5s fa_ladspa_fbdelay_60s fa_ladspa_flanger 
fa_ladspa_foldover fa_sampleformat fa_ladspa_fadDelay fa_ladspa_freeverb3 
fa_ladspa_freqTracker fa_ladspa_butthigh_iir fa_ladspa_buttlow_iir 
fa_ladspa_gsm fa_ladspa_gate fa_ladspa_giantFlange 
fa_ladspa_bandpass_a_iir fa_ladspa_bandpass_iir fa_ladspa_highpass_iir 
fa_ladspa_lowpass_iir fa_ladspa_gongBeater fa_ladspa_gong 
fa_ladspa_grain_scatter fa_ladspa_hard_gate fa_ladspa_hardLimiter 
fa_ladspa_harmonicGen fa_ladspa_hermesFilter fa_ladspa_hpf 
fa_ladspa_pitchScaleHQ fa_ladspa_identity_audio fa_ladspa_imp 
fa_ladspa_inv fa_ladspa_karaoke fa_ladspa_lcrDelay fa_ladspa_lfoPhaser 
fa_ladspa_lsFilter fa_ladspa_lofi fa_ladspa_lpf fa_ladspa_notch_iir 
fa_ladspa_matrixSpatialiser fa_ladspa_matrixMSSt fa_ladspa_matrixStMS 
fa_ladspa_mbeq fa_ladspa_multivoiceChorus fa_ladspa_pitchScale 
fa_ladspa_pointerCastDistortion fa_ladspa_rateShifter 
fa_ladspa_retroFlange fa_ladspa_revdelay fa_ladspa_ringmod_1i1o1l 
fa_ladspa_sc1 fa_ladspa_sc4 fa_ladspa_sc4m fa_ladspa_se4 fa_ladspa_sifter 
fa_ladspa_compress_peak fa_ladspa_compress_rms fa_ladspa_expand_peak 
fa_ladspa_expand_rms fa_ladspa_limit_peak fa_ladspa_limit_rms 
fa_ladspa_amp fa_ladspa_delay_c fa_ladspa_delay_l fa_ladspa_delay_n 
fa_ladspa_sine_faac fa_ladspa_sine_fcaa fa_ladspa_singlePara 
fa_ladspa_sinusWavewrapper fa_ladspa_smoothDecimate fa_ladspa_svf 
fa_ladspa_tap_autopan fa_ladspa_tap_chorusflanger fa_ladspa_tap_deesser 
fa_ladspa_tap_dynamics_m fa_ladspa_tap_dynamics_st fa_ladspa_tap_equalizer 
fa_ladspa_tap_equalizer_bw fa_ladspa_tap_doubler fa_ladspa_tap_pinknoise 
fa_ladspa_tap_pitch fa_ladspa_tap_reflector fa_ladspa_tap_reverb 
fa_ladspa_tap_rotspeak fa_ladspa_tap_limiter fa_ladspa_tap_sigmoid 
fa_ladspa_tap_stereo_echo fa_ladspa_tap_tremolo fa_ladspa_tap_tubewarmth 
fa_ladspa_tap_vibrato fa_ladspa_tapeDelay fa_ladspa_transient 
fa_ladspa_triplePara fa_ladspa_vcf303 fa_ladspa_valveRect fa_ladspa_valve 
fa_volume fa_ladspa_vynil fa_ladspa_wshape_sine fa_ladspa_shaper 
fa_ladspa_zm1

Audio Filters
.P
.B Suboptions for fa_ladspa_fourByFourPole

.P
.RS 2
.BR Frequency 1 "=<number> (1.000..20000.000, default: 5000.750)"
.P
Frequency 1
.P
.BR Feedback 1 "=<number> (-1.000..1.000, default: 5000.750)"
.P
Feedback 1
.P
.BR Frequency 2 "=<number> (1.000..20000.000, default: 10000.500)"
.P
Frequency 2
.P
.BR Feedback 2 "=<number> (-1.000..1.000, default: 10000.500)"
.P
Feedback 2
.P
.BR Frequency 3 "=<number> (1.000..20000.000, default: 15000.250)"
.P
Frequency 3
.P
.BR Feedback 3 "=<number> (-1.000..1.000, default: 15000.250)"
.P
Feedback 3
.P
.BR Frequency 4 "=<number> (1.000..20000.000, default: 20000.000)"
.P
Frequency 4
.P
.BR Feedback 4 "=<number> (-1.000..1.000, default: 20000.000)"
.P
Feedback 4
.P
.BR $run_adding "=[1|0] (default: 0)"
.P
Add effect to input data
.P
.BR $run_adding_gain "=<number> (-70.00..0.00, default: 0.00)"
.P
Add gain (dB)
.P
Overall gain for this filter. This is only valid if you add the effect
to the input data

.P
.RE
.B Suboptions for fa_ladspa_amPitchshift

.P
.RS 2
.BR Pitch shift "=<number> (0.250..4.000, default: 0.000)"
.P
Pitch shift
.P
.BR Buffer size "=<number> (1..7, default: 4)"
.P
Buffer size
.P
.BR $run_adding "=[1|0] (default: 0)"
.P
Add effect to input data
.P
.BR $run_adding_gain "=<number> (-70.00..0.00, default: 0.00)"
.P
Add gain (dB)
.P
Overall gain for this filter. This is only valid if you add the effect
to the input data

.P
.RE
.B Suboptions for fa_ladspa_alias

.P
.RS 2
.BR Aliasing level "=<number> (0.000..1.000, default: 0.000)"
.P
Aliasing level
.P
.BR $run_adding "=[1|0] (default: 0)"
.P
Add effect to input data
.P
.BR $run_adding_gain "=<number> (-70.00..0.00, default: 0.00)"
.P
Add gain (dB)
.P
Overall gain for this filter. This is only valid if you add the effect
to the input data

.P
.RE
.B Suboptions for fa_ladspa_allpass_c

.P
.RS 2
.BR Max Delay (s) "=<number> (default: 0.000)"
.P
Max Delay (s)
.P
.BR Delay Time (s) "=<number> (default: 0.000)"
.P
Delay Time (s)
.P
.BR Decay Time (s) "=<number> (default: 0.000)"
.P
Decay Time (s)
.P
.BR $run_adding "=[1|0] (default: 0)"
.P
Add effect to input data
.P
.BR $run_adding_gain "=<number> (-70.00..0.00, default: 0.00)"
.P
Add gain (dB)
.P
Overall gain for this filter. This is only valid if you add the effect
to the input data

.P
.RE
.B Suboptions for fa_ladspa_allpass_l

.P
.RS 2
.BR Max Delay (s) "=<number> (default: 0.000)"
.P
Max Delay (s)
.P
.BR Delay Time (s) "=<number> (default: 0.000)"
.P
Delay Time (s)
.P
.BR Decay Time (s) "=<number> (default: 0.000)"
.P
Decay Time (s)
.P
.BR $run_adding "=[1|0] (default: 0)"
.P
Add effect to input data
.P
.BR $run_adding_gain "=<number> (-70.00..0.00, default: 0.00)"
.P
Add gain (dB)
.P
Overall gain for this filter. This is only valid if you add the effect
to the input data

.P
.RE
.B Suboptions for fa_ladspa_allpass_n

.P
.RS 2
.BR Max Delay (s) "=<number> (default: 0.000)"
.P
Max Delay (s)
.P
.BR Delay Time (s) "=<number> (default: 0.000)"
.P
Delay Time (s)
.P
.BR Decay Time (s) "=<number> (default: 0.000)"
.P
Decay Time (s)
.P
.BR $run_adding "=[1|0] (default: 0)"
.P
Add effect to input data
.P
.BR $run_adding_gain "=<number> (-70.00..0.00, default: 0.00)"
.P
Add gain (dB)
.P
Overall gain for this filter. This is only valid if you add the effect
to the input data

.P
.RE
.B Suboptions for fa_ladspa_amp_mono

.P
.RS 2
.BR Gain "=<number> (default: 0.000)"
.P
Gain
.P
.RE
.B Suboptions for fa_ladspa_amp_stereo

.P
.RS 2
.BR Gain "=<number> (default: 0.000)"
.P
Gain
.P
.RE
.B Suboptions for fa_ladspa_artificialLatency

.P
.RS 2
.BR Delay (ms) "=<number> (0.000..10000.000, default: 2500.000)"
.P
Delay (ms)
.P
.BR $run_adding "=[1|0] (default: 0)"
.P
Add effect to input data
.P
.BR $run_adding_gain "=<number> (-70.00..0.00, default: 0.00)"
.P
Add gain (dB)
.P
Overall gain for this filter. This is only valid if you add the effect
to the input data

.P
.RE
.B Suboptions for fa_ladspa_divider

.P
.RS 2
.BR Denominator "=<number> (1..8, default: 0)"
.P
Denominator
.P
.BR $run_adding "=[1|0] (default: 0)"
.P
Add effect to input data
.P

gmerlin/ChangeLog
2007-02-12  gettextize  <bug-gnu-gettext@gnu.org>

	* m4/codeset.m4: New file, from gettext-0.15.
	* m4/gettext.m4: New file, from gettext-0.15.
	* m4/glibc2.m4: New file, from gettext-0.15.
	* m4/glibc21.m4: New file, from gettext-0.15.
	* m4/iconv.m4: New file, from gettext-0.15.
	* m4/intdiv0.m4: New file, from gettext-0.15.
	* m4/intmax.m4: New file, from gettext-0.15.
	* m4/inttypes_h.m4: New file, from gettext-0.15.
	* m4/inttypes-h.m4: New file, from gettext-0.15.
	* m4/inttypes-pri.m4: New file, from gettext-0.15.
	* m4/lcmessage.m4: New file, from gettext-0.15.
	* m4/lib-ld.m4: New file, from gettext-0.15.
	* m4/lib-link.m4: New file, from gettext-0.15.
	* m4/lib-prefix.m4: New file, from gettext-0.15.
	* m4/lock.m4: New file, from gettext-0.15.
	* m4/longdouble.m4: New file, from gettext-0.15.
	* m4/longlong.m4: New file, from gettext-0.15.
	* m4/nls.m4: New file, from gettext-0.15.
	* m4/po.m4: New file, from gettext-0.15.
	* m4/printf-posix.m4: New file, from gettext-0.15.
	* m4/progtest.m4: New file, from gettext-0.15.
	* m4/signed.m4: New file, from gettext-0.15.
	* m4/size_max.m4: New file, from gettext-0.15.
	* m4/stdint_h.m4: New file, from gettext-0.15.
	* m4/uintmax_t.m4: New file, from gettext-0.15.
	* m4/ulonglong.m4: New file, from gettext-0.15.
	* m4/visibility.m4: New file, from gettext-0.15.
	* m4/wchar_t.m4: New file, from gettext-0.15.
	* m4/wint_t.m4: New file, from gettext-0.15.
	* m4/xsize.m4: New file, from gettext-0.15.
	* m4/Makefile.am (EXTRA_DIST): Add the new files.
	* Makefile.am (SUBDIRS): Add po.
	(EXTRA_DIST): Add config.rpath.
	* configure.ac (AC_CONFIG_FILES): Add po/Makefile.in.



gmerlin/apps/cmdlinerecorder/gmerlin-record.c
/*****************************************************************
 * gmerlin - a general purpose multimedia framework and applications
 *
 * Copyright (c) 2001 - 2012 Members of the Gmerlin project
 * gmerlin-general@lists.sourceforge.net
 * http://gmerlin.sourceforge.net
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * *****************************************************************/

#include <signal.h>

#include <config.h>
#include <gmerlin/recorder.h>
#include <gmerlin/pluginregistry.h>
#include <gavl/gavl.h>
#include <gmerlin/utils.h>
#include <gmerlin/cmdline.h>
#include <gmerlin/log.h>
#include <gmerlin/translation.h>

static bg_recorder_t * recorder;

static bg_cfg_section_t * audio_section;
static bg_cfg_section_t * video_section;
static bg_cfg_section_t * metadata_section;
static bg_cfg_section_t * encoder_section;
static bg_cfg_section_t * af_section;
static bg_cfg_section_t * vf_section;
static bg_cfg_section_t * vm_section;
static bg_cfg_section_t * output_section;

const bg_parameter_info_t * audio_parameters;
const bg_parameter_info_t * video_parameters;
const bg_parameter_info_t * af_parameters;
const bg_parameter_info_t * vf_parameters;
const bg_parameter_info_t * vm_parameters;
const bg_parameter_info_t * output_parameters;

const bg_parameter_info_t * metadata_parameters;
const bg_parameter_info_t * encoder_parameters;


static bg_plugin_registry_t * plugin_reg;
static bg_cfg_registry_t * cfg_reg;

static int do_record = 0;

static int do_syslog = 0;

#define DELAY_TIME 50
#define PING_INTERVAL 20

int got_sigint = 0;

static void sigint_handler(int sig)
  {
  fprintf(stderr, "\nCTRL+C caught\n");
  got_sigint = 1;
  }

static void set_sigint_handler()
  {
  struct sigaction sa;
  sa.sa_flags = 0;
  sigemptyset(&sa.sa_mask);
  sa.sa_handler = sigint_handler;
  if (sigaction(SIGINT, &sa, NULL) == -1)
    {
    fprintf(stderr, "sigaction failed\n");
    }
  }

static void opt_aud(void * data, int * argc, char *** _argv, int arg)
  {
  if(arg >= *argc)
    {
    fprintf(stderr, "Option -aud requires an argument\n");
    exit(-1);
    }
  if(!bg_cmdline_apply_options(audio_section,
                               bg_recorder_set_audio_parameter,
                               recorder,
                               audio_parameters,
                               (*_argv)[arg]))
    exit(-1);
  bg_cmdline_remove_arg(argc, _argv, arg);
  }

static void opt_af(void * data, int * argc, char *** _argv, int arg)
  {
  if(arg >= *argc)
    {
    fprintf(stderr, "Option -af requires an argument\n");
    exit(-1);
    }
  if(!bg_cmdline_apply_options(audio_section,
                               bg_recorder_set_audio_filter_parameter,
                               recorder,
                               af_parameters,
                               (*_argv)[arg]))
    exit(-1);
  bg_cmdline_remove_arg(argc, _argv, arg);
  }

static void opt_vid(void * data, int * argc, char *** _argv, int arg)
  {
  if(arg >= *argc)
    {
    fprintf(stderr, "Option -vid requires an argument\n");
    exit(-1);
    }
  if(!bg_cmdline_apply_options(video_section,
                               bg_recorder_set_video_parameter,
                               recorder,
                               video_parameters,
                               (*_argv)[arg]))
    exit(-1);
  bg_cmdline_remove_arg(argc, _argv, arg);
  }

static void opt_vf(void * data, int * argc, char *** _argv, int arg)
  {
  if(arg >= *argc)
    {
    fprintf(stderr, "Option -vf requires an argument\n");
    exit(-1);
    }
  if(!bg_cmdline_apply_options(video_section,
                               bg_recorder_set_video_filter_parameter,
                               recorder,
                               vf_parameters,
                               (*_argv)[arg]))
    exit(-1);
  bg_cmdline_remove_arg(argc, _argv, arg);
  }

static void opt_vm(void * data, int * argc, char *** _argv, int arg)
  {
  if(arg >= *argc)
    {
    fprintf(stderr, "Option -vm requires an argument\n");
    exit(-1);
    }
  if(!bg_cmdline_apply_options(video_section,
                               bg_recorder_set_video_monitor_parameter,
                               recorder,
                               vm_parameters,
                               (*_argv)[arg]))
    exit(-1);
  bg_cmdline_remove_arg(argc, _argv, arg);
  }

static void opt_m(void * data, int * argc, char *** _argv, int arg)
  {
  if(arg >= *argc)
    {
    fprintf(stderr, "Option -m requires an argument\n");
    exit(-1);
    }
  if(!bg_cmdline_apply_options(metadata_section,
                               bg_recorder_set_metadata_parameter,
                               recorder,
                               metadata_parameters,
                               (*_argv)[arg]))
    exit(-1);
  bg_cmdline_remove_arg(argc, _argv, arg);
  }

static void opt_r(void * data, int * argc, char *** _argv, int arg)
  {
  do_record = 1;
  }

static void opt_enc(void * data, int * argc, char *** _argv, int arg)
  {
  if(arg >= *argc)
    {
    fprintf(stderr, "Option -enc requires an argument\n");
    exit(-1);
    }
  if(!bg_cmdline_apply_options(encoder_section,
                               /* bg_recorder_set_metadata_parameter */ NULL,
                               /* recorder */ NULL,
                               encoder_parameters,
                               (*_argv)[arg]))
    exit(-1);
  bg_cmdline_remove_arg(argc, _argv, arg);
  }

static void opt_o(void * data, int * argc, char *** _argv, int arg)
  {
  if(arg >= *argc)
    {
    fprintf(stderr, "Option -o requires an argument\n");
    exit(-1);
    }
  if(!bg_cmdline_apply_options(encoder_section,
                               bg_recorder_set_output_parameter,
                               recorder,
                               output_parameters,
                               (*_argv)[arg]))
    exit(-1);
  bg_cmdline_remove_arg(argc, _argv, arg);
  }

static void opt_syslog(void * data, int * argc, char *** _argv, int arg)
  {
  if(arg >= *argc)
    {
    fprintf(stderr, "Option -syslog requires an argument\n");
    exit(-1);
    }
  bg_log_syslog_init((*_argv)[arg]);
  do_syslog = 1;
  bg_cmdline_remove_arg(argc, _argv, arg);
  }

static bg_cmdline_arg_t global_options[] =
  {
    {
      .arg =         "-aud",
      .help_arg =    "<audio_options>",
      .help_string = "Set audio recording options",
      .callback =    opt_aud,
    },
    {
      .arg =         "-vid",
      .help_arg =    "<video_options>",
      .help_string = "Set video recording options",
      .callback =    opt_vid,
    },
    {
      .arg =         "-af",
      .help_arg =    "<audio filter options>",
      .help_string = "Set audio filter options",
      .callback =    opt_af,
    },
    {
      .arg =         "-vf",
      .help_arg =    "<video filter options>",
      .help_string = "Set video filter options",
      .callback =    opt_vf,
    },
    {
      .arg =         "-vm",
      .help_arg =    "<video monitor options>",
      .help_string = "Set video monitor options",
      .callback =    opt_vm,
    },
    {
      .arg =         "-m",
      .help_arg =    "<metadata_options>",
      .help_string = "Set metadata options",
      .callback =    opt_m,
    },
    {
      .arg =         "-enc",
      .help_arg =    "<encoding_options>",
      .help_string = "Set encoding options",
      .callback =    opt_enc,
      
    },
    {
      .arg =         "-o",
      .help_arg =    "<output_options>",
      .help_string = "Set output options",
      .callback =    opt_o,
      
    },
    {
      .arg =         "-r",
      .help_string = "Record to file",
      .callback =    opt_r,
    },
    {
      .arg =         "-syslog",
      .help_arg =    "<name>",
      .help_string = "Set log messages to syslog",
      .callback =    opt_syslog,
    },
    {
      /* End */
    }
  };

static void update_global_options()
  {
  global_options[0].parameters = audio_parameters;
  global_options[1].parameters = video_parameters;
  global_options[2].parameters = af_parameters;
  global_options[3].parameters = vf_parameters;
  global_options[4].parameters = vm_parameters;
  global_options[5].parameters = metadata_parameters;
  global_options[6].parameters = encoder_parameters;
  global_options[7].parameters = output_parameters;
  }

const bg_cmdline_app_data_t app_data =
  {
    .package =  PACKAGE,
    .version =  VERSION,
    .name =     "gmerlin-record",
    .long_name = TRS("Commandline recorder"),
    .synopsis = TRS("[options]\n"),
    .help_before = TRS("Commandline recorder\n"),
    .args = (bg_cmdline_arg_array_t[]) { { TRS("Options"), global_options },
                                       {  } },
    .files = (bg_cmdline_ext_doc_t[])
    { { "~/.gmerlin/plugins.xml",
        TRS("Cache of the plugin registry (shared by all applicatons)") },
      { "~/.gmerlin/generic/config.xml",
        TRS("Default plugin parameters are read from there. Use gmerlin_plugincfg to change them.") },
      { /* End */ }
    },
    
  };


  
int main(int argc, char ** argv)
  {
  bg_cfg_section_t * cfg_section;
  char * tmp_path;
  int timeout_counter = 0;

  gavl_time_t delay_time = DELAY_TIME * 1000;

  set_sigint_handler();
  
  /* We must initialize the random number generator if we want the
     Vorbis encoder to work */
  srand(time(NULL));
  
  /* Create plugin regsitry */
  cfg_reg = bg_cfg_registry_create();
  tmp_path =  bg_search_file_read("generic", "config.xml");
  bg_cfg_registry_load(cfg_reg, tmp_path);
  if(tmp_path)
    free(tmp_path);
  
  cfg_section = bg_cfg_registry_find_section(cfg_reg, "plugins");
  plugin_reg = bg_plugin_registry_create(cfg_section);

  recorder = bg_recorder_create(plugin_reg);
  
  /* Create config sections */
  
  audio_parameters = bg_recorder_get_audio_parameters(recorder);
  audio_section =
    bg_cfg_section_create_from_parameters("audio", audio_parameters);

  video_parameters = bg_recorder_get_video_parameters(recorder);
  video_section =
    bg_cfg_section_create_from_parameters("video", video_parameters);

  af_parameters = bg_recorder_get_audio_filter_parameters(recorder);
  af_section =
    bg_cfg_section_create_from_parameters("audio_filters", af_parameters);

  vf_parameters = bg_recorder_get_video_filter_parameters(recorder);
  vf_section =
    bg_cfg_section_create_from_parameters("video_filters", vf_parameters);

  vm_parameters = bg_recorder_get_video_monitor_parameters(recorder);
  vm_section =
    bg_cfg_section_create_from_parameters("video_monitor", vm_parameters);

  metadata_parameters = bg_recorder_get_metadata_parameters(recorder);
  metadata_section =
    bg_cfg_section_create_from_parameters("metadata", metadata_parameters);

  encoder_parameters = bg_recorder_get_encoder_parameters(recorder);
  
  encoder_section =
    bg_cfg_section_create_from_parameters("encoder", encoder_parameters);

  output_parameters = bg_recorder_get_output_parameters(recorder);
  output_section =
    bg_cfg_section_create_from_parameters("output", output_parameters);
  
  update_global_options();

  /* Apply default options */
  
  bg_cfg_section_apply(audio_section, audio_parameters,
                       bg_recorder_set_audio_parameter, recorder);
  bg_cfg_section_apply(video_section, video_parameters,
                       bg_recorder_set_video_parameter, recorder);
  bg_cfg_section_apply(af_section, af_parameters,
                       bg_recorder_set_audio_filter_parameter, recorder);
  bg_cfg_section_apply(vf_section, vf_parameters,
                       bg_recorder_set_video_filter_parameter, recorder);
  bg_cfg_section_apply(vm_section, vm_parameters,
                       bg_recorder_set_video_monitor_parameter, recorder);
  bg_cfg_section_apply(metadata_section, metadata_parameters,
                       bg_recorder_set_metadata_parameter, recorder);
  bg_cfg_section_apply(output_section, output_parameters,
                       bg_recorder_set_output_parameter, recorder);
  
  /* Get commandline options */
  bg_cmdline_init(&app_data);
  bg_cmdline_parse(global_options, &argc, &argv, NULL);

  /* Set encoding options */
  bg_recorder_set_encoder_section(recorder, encoder_section);
  
  bg_recorder_record(recorder, do_record);
  bg_recorder_run(recorder);
  
  while(1)
    {
    if(!timeout_counter)
      {
      if(!bg_recorder_ping(recorder))
        break;
      }

    if(got_sigint)
      break;
    
    timeout_counter++;
    if(timeout_counter >= PING_INTERVAL)
      timeout_counter = 0;

    if(do_syslog)
      bg_log_syslog_flush();
    
    gavl_time_delay(&delay_time);
    }
  
  bg_recorder_destroy(recorder);
  return 0;
  }
